# Preprocessing

In [1]:
import pandas as pd
import numpy as np

In [2]:
from functions import *

### Import the datasets

In [3]:
# Open the raw datasets
# Heavy Metal
hm_df = pd.read_excel('data/AllHMFiles.xlsx', engine='openpyxl')

# Pesticides
pesticides_df = pd.read_excel('data/AllPesticidesFiles.xlsx', engine='openpyxl')
classe_pesticides_df = pd.read_excel('data/pesticides.xlsx', engine='openpyxl')

# Surfaces
surfaces_df = pd.read_excel('data/absSurfs.xlsx', engine='openpyxl')

# Distances
distances_df = pd.read_excel('data/distsOneSheet.xlsx', engine='openpyxl')

In [4]:
hm_df.head(5)

REF....SUBSTANCE   PERIOD   Lead  Cadmium  Arsenic  Mercury  Copper  \
0            B0017  2017/P1  0.081     0.81      0.0      0.0     5.5   
1            B0019  2017/P1  0.062     0.36      0.0      0.0     8.5   
2            B0021  2017/P1  0.150     0.61      0.0      0.0     7.0   
3            B0023  2017/P1  0.067     0.35      0.0      0.0     6.9   
4            B0037  2017/P1  0.120     0.07      0.0      0.0     5.6   

   Chromium  Zinc  Chrome  
0     0.032    40     NaN  
1     0.029    33     NaN  
2     0.038    37     NaN  
3     0.031    37     NaN  
4     0.052    35     NaN

In [5]:
pesticides_df.head(5)

REF....SUBSTANCE   PERIOD  Aldrin  Benalaxyl  Chlorpropham  Cyanofenphos  \
0            B0017  2017/P1     0.0     0.0000          0.00           0.0   
1            B0019  2017/P1     0.0     0.0188          0.00           0.0   
2            B0021  2017/P1     0.0     0.0711          0.00           0.0   
3            B0023  2017/P1     0.0     0.0294          0.00           0.0   
4            B0037  2017/P1     0.0     0.0000          0.01           0.0   

   Cypermethrin  Ethofumesate  Fludioxonil  Fluopyram  ...  Nicosulfuron  \
0           0.0           0.0          0.0        0.0  ...           NaN   
1           0.0           0.0          0.0        0.0  ...           NaN   
2           0.0           0.0          0.0        0.0  ...           NaN   
3           0.0           0.0          0.0        0.0  ...           NaN   
4           0.0           0.0          0.0        0.0  ...           NaN   

   Piperonylbutoxide  Prosulfuron  Simazine  Terbuthylazine  Dieldrin  \
0                NaN          NaN       NaN             NaN       NaN   
1                NaN          NaN       NaN             NaN       NaN   
2                NaN          NaN       NaN             NaN       NaN   
3                NaN          NaN       NaN             NaN       NaN   
4                NaN          NaN       NaN             NaN       NaN   

   Fenamidone  Trifluralin  Cycloate  Pentachloroanisol  
0         NaN          NaN       NaN                NaN  
1         NaN          NaN       NaN                NaN  
2         NaN          NaN       NaN                NaN  
3         NaN          NaN       NaN                NaN  
4         NaN          NaN       NaN                NaN  

[5 rows x 117 columns]

In [6]:
classe_pesticides_df.head(5)

importName           name    statusEN   LMR  \
0      Aclonifen      aclonifen  authorized  0.05   
1         Aldrin         aldrin  prohibited  0.01   
2   Ametoctradin   ametoctradin  authorized  0.05   
3  Anthraquinone  anthraquinone  prohibited  0.02   
4   Azoxystrobin   azoxystrobin  authorized  0.05   

                          familyEN          typeEN  
0                   diphenyl ether       herbicide  
1                   organochlorine     insecticide  
2               triazolopyrimidine       fungicide  
3  polycyclic aromatic hydrocarbon  bird repellant  
4                      strobilurin       fungicide

In [7]:
surfaces_df.head(5)

Site  111           112           121            122  123  124  131  132  \
0  F0069  0.0  1.064122e+06  3.587087e+05       0.000000  0.0  0.0  0.0    0   
1  F0025  0.0  0.000000e+00  0.000000e+00  412752.692383  0.0  0.0  0.0    0   
2  F0027  0.0  2.741746e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
3  B0128  0.0  2.470089e+06  4.053755e+06  497496.251770  0.0  0.0  0.0    0   
4  F0023  0.0  0.000000e+00  2.559254e+05       0.000000  0.0  0.0  0.0    0   

   133  ...  411  412  421  422  423  511            512  521  522  523  
0  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
1  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
2  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
3  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
4  0.0  ...    0  0.0    0    0  0.0  0.0  284728.392578    0    0  0.0  

[5 rows x 45 columns]

In [8]:
distances_df.head(5)

Site     polyID  classCLC         dist
0  F0069   FR-88090       231  1183.022362
1  F0069  FR-131773       242   289.926324
2  F0069   FR-46167       211   221.344153
3  F0069   FR-11047       112  1038.383069
4  F0069  FR-195343       311   724.045683

### Heavy Metal

### Pesticides

Fungicide :

In [9]:
classe_pesticides_df[classe_pesticides_df['typeEN'] == 'fungicide'].head(5)

importName                       name    statusEN   LMR  \
2         Ametoctradin               ametoctradin  authorized  0.05   
4         Azoxystrobin               azoxystrobin  authorized  0.05   
5            Benalaxyl                  benalaxyl  prohibited  0.05   
7  Benthiavalicarb.iso  benthiavalicarb-isopropyl  authorized  0.05   
8     Benzovindiflupyr           benzovindiflupyr  authorized  0.05   

             familyEN     typeEN  
2  triazolopyrimidine  fungicide  
4         strobilurin  fungicide  
5        phenylamides  fungicide  
7           carbamate  fungicide  
8               amide  fungicide

In [10]:
fungicides = classe_pesticides_df[classe_pesticides_df['typeEN'] == 'fungicide'].importName.to_list()
fungicides

['Ametoctradin',
 'Azoxystrobin',
 'Benalaxyl',
 'Benthiavalicarb.iso',
 'Benzovindiflupyr',
 'Boscalid',
 'Bupirimate',
 'Captan',
 'Carbendazim',
 'Chlorothalonil',
 'Cyazofamid',
 'Cyflufenamid',
 'Cymoxanil',
 'Cyprodinil',
 'Difenoconazole',
 'Dimethomorph',
 'Dimoxystrobin',
 'Dodine',
 'Epoxiconazole',
 'Ethirimol',
 'Famoxadone',
 'Fenamidone',
 'Fenbuconazole',
 'Fenhexamid',
 'Fenpropidin',
 'Fenpropimorph',
 'Fluazinam',
 'Fludioxonil',
 'Fluopicolide',
 'Fluopyram',
 'Flutolanil',
 'Fluxapyroxad',
 'Hexaconazole',
 'Iprovalicarb',
 'Isofetamid',
 'Mandipropamid',
 'Metalaxyl',
 'Metconazole',
 'Metrafenone',
 'Penconazole',
 'Pencycuron',
 'Pentachloroanisol',
 'Prochloraz',
 'Propamocarb',
 'Propiconazole',
 'Prothioconazole_d',
 'Pyraclostrobin',
 'Pyrimethanil',
 'Quinoxyfen',
 'Spiroxamine',
 'Tetraconazole',
 'Thiabendazole',
 'Thiophanate.methyl',
 'Trifloxystrobin',
 'X2.phenylphenol',
 'Zoxamide']

Insecticide :

In [11]:
classe_pesticides_df[classe_pesticides_df['typeEN'] == 'insecticide'].head(5)

importName                 name          statusEN   LMR  \
1                Aldrin               aldrin        prohibited  0.01   
13  Chlorantraniliprole  chlorantraniliprole        authorized  0.05   
17  Chlorpyrifos.methyl  chlorpyrifos-methyl        prohibited  0.05   
19            Coumaphos            coumaphos        prohibited  0.10   
20         Cyanofenphos         cyanofenphos  not yet assessed   NaN   

               familyEN       typeEN  
1        organochlorine  insecticide  
13  anthranilic diamide  insecticide  
17      organophosphate  insecticide  
19        thiophosphate  insecticide  
20      organophosphate  insecticide

In [12]:
insecticides = classe_pesticides_df[classe_pesticides_df['typeEN'] == 'insecticide'].importName.to_list()
insecticides

['Aldrin',
 'Chlorantraniliprole',
 'Chlorpyrifos.methyl',
 'Coumaphos',
 'Cyanofenphos',
 'Cyflumetofen',
 'Cypermethrin',
 'Deet',
 'Dieldrin',
 'Etofenprox',
 'Fenoxycarb',
 'Fenpropathrin',
 'Flonicamid',
 'Flucythrinate',
 'Flupyradifurone',
 'Imidacloprid',
 'Indoxacarb',
 'Methoxyfenozide',
 'Phosmet',
 'Piperonylbutoxide',
 'Pirimicarb',
 'Spinosad',
 'Spirotetramat',
 'Tau.fluvalinate',
 'Tebufenozide',
 'Tetradifon',
 'Thiacloprid']

Herbicide : 

In [13]:
classe_pesticides_df[classe_pesticides_df['typeEN'] == 'herbicide'].head(5)

importName           name    statusEN   LMR          familyEN  \
0       Aclonifen      aclonifen  authorized  0.05    diphenyl ether   
6     Benfluralin    benfluralin  authorized  0.05    dinitroaniline   
15  Chlorotoluron  chlorotoluron  authorized  0.05        phenylurea   
16   Chlorpropham   chlorpropham  prohibited  0.05         carbamate   
18      Clethodim      clethodim  authorized  0.05  cyclohexanedione   

       typeEN  
0   herbicide  
6   herbicide  
15  herbicide  
16  herbicide  
18  herbicide

In [14]:
herbicides = classe_pesticides_df[classe_pesticides_df['typeEN'] == 'herbicide'].importName.to_list()
herbicides

['Aclonifen',
 'Benfluralin',
 'Chlorotoluron',
 'Chlorpropham',
 'Clethodim',
 'Cycloate',
 'Diflufenican',
 'Ethofumesate',
 'Fluazifop',
 'Flufenacet',
 'Metamitron',
 'Methabenzthiazuron',
 'Metobromuron',
 'Metolachlor',
 'Nicosulfuron',
 'Pendimethalin',
 'Phenmedipham',
 'Propanil',
 'Propyzamide',
 'Prosulfocarb',
 'Prosulfuron',
 'Simazine',
 'Terbuthylazine',
 'Tri.allate',
 'Trifluralin']

Others :

In [15]:
classe_pesticides_df[(classe_pesticides_df['typeEN'] != 'fungicide') & 
                     (classe_pesticides_df['typeEN'] != 'insecticide') & 
                     (classe_pesticides_df['typeEN'] != 'herbicide')]

importName           name    statusEN   LMR  \
3           Anthraquinone  anthraquinone  prohibited  0.02   
34          Diphenylamine  diphenylamine  prohibited  0.05   
69       Methiocarb_total     methiocarb  prohibited  0.05   
86             Propargite     propargite  prohibited  0.05   
100          Tebuconazole   tebuconazole  authorized  0.05   
112  X6.benzylaminopurine  benzyladenine  authorized   NaN   
114                   NaN            NaN         NaN   NaN   

                            familyEN  \
3    polycyclic aromatic hydrocarbon   
34                             amine   
69                         carbamate   
86                               NaN   
100                         triazole   
112                       cytokinins   
114                              NaN   

                                           typeEN  
3                                  bird repellant  
34    growth regulator, fungicide and insecticide  
69   insecticide, molluscicide and bird repellent  
86                                      acaricide  
100                fungicide and growth regulator  
112                              growth regulator  
114                                           NaN

##### Choice to treat the others

Adding to the corresponding categories, stay in "other" ? What about the growth regulator ?

#### Add some columns

In [16]:
pesticides_df['qtt_fungicides'] = pesticides_df.apply(qtt_pesticides, args = fungicides, axis = 1)
pesticides_df['qtt_insecticides'] = pesticides_df.apply(qtt_pesticides, args = insecticides, axis = 1)
pesticides_df['qtt_herbicides'] = pesticides_df.apply(qtt_pesticides, args = herbicides, axis = 1)

In [17]:
pesticides_df['nb_fungicides'] = pesticides_df.apply(nb_pesticides, args = fungicides, axis = 1)
pesticides_df['nb_insecticides'] = pesticides_df.apply(nb_pesticides, args = insecticides, axis = 1)
pesticides_df['nb_herbicides'] = pesticides_df.apply(nb_pesticides, args = herbicides, axis = 1)

In [18]:
pesticides_df['lmr_fungicides'] = pesticides_df.apply(lmr_pesticides, args = fungicides, axis = 1)
pesticides_df['lmr_insecticides'] = pesticides_df.apply(lmr_pesticides, args = insecticides, axis = 1)
pesticides_df['lmr_herbicides'] = pesticides_df.apply(lmr_pesticides, args = herbicides, axis = 1)

In [19]:
pesticides_df.head(10)

REF....SUBSTANCE   PERIOD  Aldrin  Benalaxyl  Chlorpropham  Cyanofenphos  \
0            B0017  2017/P1     0.0     0.0000          0.00           0.0   
1            B0019  2017/P1     0.0     0.0188          0.00           0.0   
2            B0021  2017/P1     0.0     0.0711          0.00           0.0   
3            B0023  2017/P1     0.0     0.0294          0.00           0.0   
4            B0037  2017/P1     0.0     0.0000          0.01           0.0   
5            B0039  2017/P1     0.0     0.0000          0.00           0.0   
6            B0041  2017/P1     0.0     0.0000          0.00           0.0   
7            B0043  2017/P1     0.0     0.0000          0.00           0.0   
8            B0045  2017/P1     0.0     0.0000          0.00           0.0   
9            B0059  2017/P1     0.0     0.0000          0.00           0.0   

   Cypermethrin  Ethofumesate  Fludioxonil  Fluopyram  ...  Pentachloroanisol  \
0          0.00           0.0         0.00        0.0  ...                NaN   
1          0.00           0.0         0.00        0.0  ...                NaN   
2          0.00           0.0         0.00        0.0  ...                NaN   
3          0.00           0.0         0.00        0.0  ...                NaN   
4          0.00           0.0         0.00        0.0  ...                NaN   
5          0.01           0.0         0.00        0.0  ...                NaN   
6          0.00           0.0         0.00        0.0  ...                NaN   
7          0.00           0.0         0.01        0.0  ...                NaN   
8          0.00           0.0         0.00        0.0  ...                NaN   
9          0.00           0.0         0.00        0.0  ...                NaN   

   qtt_fungicides  qtt_insecticides  qtt_herbicides  nb_fungicides  \
0          0.0000            0.0000            0.00              0   
1          0.0188            0.0000            0.00              1   
2          0.0711            0.0000            0.00              1   
3          0.0294            0.0000            0.00              1   
4          0.0000            0.0000            0.01              0   
5          0.0000            0.0208            0.00              0   
6          0.0000            0.0000            0.00              0   
7          0.0100            0.0000            0.00              1   
8          0.0000            0.0000            0.00              0   
9          0.0145            0.0000            0.00              1   

   nb_insecticides  nb_herbicides  lmr_fungicides  lmr_insecticides  \
0                0              0               0                 0   
1                0              0               0                 0   
2                0              0               1                 0   
3                0              0               0                 0   
4                0              1               0                 0   
5                2              0               0                 0   
6                0              0               0                 0   
7                0              0               0                 0   
8                0              0               0                 0   
9                0              0               0                 0   

   lmr_herbicides  
0               0  
1               0  
2               0  
3               0  
4               0  
5               0  
6               0  
7               0  
8               0  
9               0  

[10 rows x 126 columns]

In [20]:
pesticides_df['year'] = pesticides_df['PERIOD'].apply(lambda x: x[:4])
pesticides_df['trimester'] = pesticides_df['PERIOD'].apply(lambda x: x[-1:])
pesticides_df.sample(10)

REF....SUBSTANCE   PERIOD  Aldrin  Benalaxyl  Chlorpropham  Cyanofenphos  \
248            B0093  2019/P2     NaN     0.0000           NaN           NaN   
412            F0009  2020/P2     NaN     0.0000           NaN           NaN   
276           FE0001  2019/P2     NaN     0.0000           NaN           NaN   
130            F0011  2018/P2     NaN        NaN           NaN           NaN   
11             B0067  2017/P1     0.0     0.0000           0.0           0.0   
346            B0043  2020/P1     NaN     0.0000           0.0           NaN   
268            F0025  2019/P2     NaN     0.0000           NaN           NaN   
164            F0007  2018/P3     NaN     0.0101           NaN           NaN   
246            B0061  2019/P2     NaN     0.0152           NaN           NaN   
137            F0023  2018/P2     NaN        NaN           NaN           NaN   

     Cypermethrin  Ethofumesate  Fludioxonil  Fluopyram  ...  \
248           NaN           NaN          0.0       0.00  ...   
412           NaN           NaN          0.0       0.00  ...   
276           NaN           NaN          0.0       0.00  ...   
130           NaN           NaN          0.0       0.00  ...   
11            0.0           0.0          0.0       0.01  ...   
346           NaN           0.0          0.0       0.00  ...   
268           NaN           NaN          0.0       0.00  ...   
164           NaN           NaN          0.0        NaN  ...   
246           NaN           NaN          0.0       0.00  ...   
137           NaN           NaN          0.0       0.00  ...   

     qtt_insecticides  qtt_herbicides  nb_fungicides  nb_insecticides  \
248               0.0          0.0069              0                0   
412               0.0          0.0000              0                0   
276               0.0          0.0000              0                0   
130               0.0          0.0000              0                0   
11                0.0          0.0000              1                0   
346               0.0          0.0620              3                0   
268               0.0          0.0000              0                0   
164               0.0          0.0000              1                0   
246               0.0          0.0000              1                0   
137               0.0          0.0000              0                0   

     nb_herbicides  lmr_fungicides  lmr_insecticides  lmr_herbicides  year  \
248              1               0                 0               0  2019   
412              0               0                 0               0  2020   
276              0               0                 0               0  2019   
130              0               0                 0               0  2018   
11               0               0                 0               0  2017   
346              2               1                 0               1  2020   
268              0               0                 0               0  2019   
164              0               0                 0               0  2018   
246              0               0                 0               0  2019   
137              0               0                 0               0  2018   

     trimester  
248          2  
412          2  
276          2  
130          2  
11           1  
346          1  
268          2  
164          3  
246          2  
137          2  

[10 rows x 128 columns]

##### Alternate (smaller) dataset for pesticides

In [21]:
pesti_names = pesticides_df.columns.to_list()[1:-11]

In [22]:
small_pesticides_df = pesticides_df.drop(pesti_names, axis=1)
small_pesticides_df.head(5)

REF....SUBSTANCE  qtt_fungicides  qtt_insecticides  qtt_herbicides  \
0            B0017          0.0000               0.0            0.00   
1            B0019          0.0188               0.0            0.00   
2            B0021          0.0711               0.0            0.00   
3            B0023          0.0294               0.0            0.00   
4            B0037          0.0000               0.0            0.01   

   nb_fungicides  nb_insecticides  nb_herbicides  lmr_fungicides  \
0              0                0              0               0   
1              1                0              0               0   
2              1                0              0               1   
3              1                0              0               0   
4              0                0              1               0   

   lmr_insecticides  lmr_herbicides  year trimester  
0                 0               0  2017         1  
1                 0               0  2017         1  
2                 0               0  2017         1  
3                 0               0  2017         1  
4                 0               0  2017         1

##### Alternate (group by) dataset for pesticides

In [23]:
group_pesti_df = pesticides_df.groupby('REF....SUBSTANCE').agg(max)

In [24]:
group_pesti_df = group_pesti_df.loc[:,'lmr_fungicides' : 'lmr_herbicides']
group_pesti_df

lmr_fungicides  lmr_insecticides  lmr_herbicides
REF....SUBSTANCE                                                  
B0012                          1                 0               0
B0017                          1                 0               0
B0019                          0                 0               0
B0021                          1                 0               0
B0023                          0                 0               0
B0032                          1                 0               0
B0037                          1                 0               0
B0039                          1                 0               0
B0041                          1                 0               0
B0043                          1                 0               1
B0045                          1                 0               0
B0059                          1                 0               0
B0061                          0                 0               0
B0067                          1                 0               0
B0069                          1                 1               0
B0077                          1                 1               0
B0093                          1                 0               0
B0096                          0                 0               0
B0103                          1                 0               0
B0105                          1                 1               1
B0116                          0                 0               0
B0128                          0                 0               0
B0140                          0                 0               0
B0142                          1                 0               0
B0152                          1                 0               0
B0158                          1                 0               0
B0163                          1                 0               0
F0001                          1                 1               0
F0003                          1                 0               0
F0005                          0                 0               0
F0007                          0                 0               0
F0009                          0                 0               0
F0011                          0                 1               0
F0013                          0                 0               0
F0015                          1                 0               0
F0017                          0                 0               0
F0019                          1                 0               0
F0020                          1                 0               0
F0021                          1                 0               0
F0023                          0                 0               1
F0025                          0                 0               0
F0029                          1                 0               0
F0031                          0                 0               0
F0033                          0                 0               0
F0035                          0                 0               1
F0039                          1                 0               0
F0041                          0                 0               0
F0043                          0                 0               0
F0055                          0                 0               0
F0057                          0                 0               0
F0061                          1                 0               0
F0063                          1                 0               0
F0065                          0                 0               0
F0069                          0                 0               1
F0071                          1                 0               0
F0073                          1                 1               0
FE0001                         0                 0               0

### Sites

##### Getting the area for each zone

In [25]:
distances_df.head(5)

Site     polyID  classCLC         dist
0  F0069   FR-88090       231  1183.022362
1  F0069  FR-131773       242   289.926324
2  F0069   FR-46167       211   221.344153
3  F0069   FR-11047       112  1038.383069
4  F0069  FR-195343       311   724.045683

In [26]:
surfaces_df.head(5)

Site  111           112           121            122  123  124  131  132  \
0  F0069  0.0  1.064122e+06  3.587087e+05       0.000000  0.0  0.0  0.0    0   
1  F0025  0.0  0.000000e+00  0.000000e+00  412752.692383  0.0  0.0  0.0    0   
2  F0027  0.0  2.741746e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
3  B0128  0.0  2.470089e+06  4.053755e+06  497496.251770  0.0  0.0  0.0    0   
4  F0023  0.0  0.000000e+00  2.559254e+05       0.000000  0.0  0.0  0.0    0   

   133  ...  411  412  421  422  423  511            512  521  522  523  
0  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
1  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
2  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
3  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
4  0.0  ...    0  0.0    0    0  0.0  0.0  284728.392578    0    0  0.0  

[5 rows x 45 columns]

In [27]:
# surfaces_df['country'] = surfaces_df['Site'].apply(lambda x: 'FR' if x.startswith('F') else 'BE')
# surfaces_df['country'].value_counts()

In [28]:
sites_df = distances_df

In [29]:
sites_df['surface'] = sites_df.apply(get_surface, axis = 1)

In [30]:
sites_df.sample(10)

Site     polyID  classCLC         dist       surface
493  F0059  FR-206553       311   455.396503  6.676108e+05
504  F0061   FR-18674       112   604.984583  2.735732e+06
477  B0032    BE-6450       211   452.897571  2.012516e+06
241  B0043   BE-24840       511   669.864247  2.367046e+05
182  B0163    BE-4481       122  1295.260111  2.477849e+04
85   F0001   FR-15763       112  1123.831066  4.740270e+05
434  B0105    BE-1647       121     4.419787  5.098306e+05
445  B0103    BE-8293       211   714.817510  3.931007e+05
465  F0039  FR-200725       311   457.842778  5.577001e+05
489  F0045   FR-35466       142     0.000000  5.368495e+06

In [31]:
sites_df['country'] = sites_df['polyID'].apply(lambda x: 'FR' if x.startswith('FR') else 'BE')
sites_df.sample(10)

Site     polyID  classCLC         dist       surface country
267  B0059    BE-4526       122   551.599486  2.585516e+05      BE
609  B0152    BE-6970       211  1451.428316  1.121794e+06      BE
112  B0012    BE-2861       112  1126.163642  3.756405e+06      BE
625  B0158    BE-5811       211  1003.204616  3.326810e+06      BE
47   F0071  FR-254702       324   634.579612  7.933020e+05      FR
551  F0005  FR-121947       242  1084.942589  2.728893e+05      FR
100  F0037   FR-31037       123   709.411950  1.037133e+06      FR
408  B0017   BE-23796       313  1445.070443  1.871540e+06      BE
442  B0103    BE-3873       112  1053.448569  4.637138e+05      BE
589  F0007  FR-187751       311    37.466608  3.211653e+06      FR

In [32]:
sites_df['country'].value_counts()

FR    335
BE    297
Name: country, dtype: int64

In [33]:
sites_df = sites_df.drop(['polyID'], axis=1)
sites_df.sample(5)

Site  classCLC         dist       surface country
475  B0032       121     0.000000  1.005540e+06      BE
479  B0032       211  1257.131953  2.012516e+06      BE
142  F0041       121     0.000000  3.383666e+06      FR
285  B0093       211   346.401772  2.350786e+06      BE
302  F0055       324   206.905784  9.163058e+05      FR

##### Get one dataset with : beehives and area/distance for each zone

In [34]:
surfaces_df

Site  111           112           121            122  123  124  131  132  \
0   F0069  0.0  1.064122e+06  3.587087e+05       0.000000  0.0  0.0  0.0    0   
1   F0025  0.0  0.000000e+00  0.000000e+00  412752.692383  0.0  0.0  0.0    0   
2   F0027  0.0  2.741746e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
3   B0128  0.0  2.470089e+06  4.053755e+06  497496.251770  0.0  0.0  0.0    0   
4   F0023  0.0  0.000000e+00  2.559254e+05       0.000000  0.0  0.0  0.0    0   
..    ...  ...           ...           ...            ...  ...  ...  ...  ...   
58  F0007  0.0  0.000000e+00  0.000000e+00       0.000000  0.0  0.0  0.0    0   
59  B0069  0.0  8.704243e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
60  B0152  0.0  4.575279e+06  5.726391e+05  201434.000214  0.0  0.0  0.0    0   
61  F0003  0.0  3.425116e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
62  B0158  0.0  1.467398e+05  0.000000e+00  318178.187439  0.0  0.0  0.0    0   

    133  ...  411  412  421  422  423  511            512  521  522  523  
0   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
1   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
2   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
3   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
4   0.0  ...    0  0.0    0    0  0.0  0.0  284728.392578    0    0  0.0  
..  ...  ...  ...  ...  ...  ...  ...  ...            ...  ...  ...  ...  
58  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
59  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
60  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
61  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
62  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  

[63 rows x 45 columns]

In [35]:
classCLC = surfaces_df.columns.to_list()[1:]

In [36]:
by_sites_df = surfaces_df.copy(deep=True)
by_sites_df

Site  111           112           121            122  123  124  131  132  \
0   F0069  0.0  1.064122e+06  3.587087e+05       0.000000  0.0  0.0  0.0    0   
1   F0025  0.0  0.000000e+00  0.000000e+00  412752.692383  0.0  0.0  0.0    0   
2   F0027  0.0  2.741746e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
3   B0128  0.0  2.470089e+06  4.053755e+06  497496.251770  0.0  0.0  0.0    0   
4   F0023  0.0  0.000000e+00  2.559254e+05       0.000000  0.0  0.0  0.0    0   
..    ...  ...           ...           ...            ...  ...  ...  ...  ...   
58  F0007  0.0  0.000000e+00  0.000000e+00       0.000000  0.0  0.0  0.0    0   
59  B0069  0.0  8.704243e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
60  B0152  0.0  4.575279e+06  5.726391e+05  201434.000214  0.0  0.0  0.0    0   
61  F0003  0.0  3.425116e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
62  B0158  0.0  1.467398e+05  0.000000e+00  318178.187439  0.0  0.0  0.0    0   

    133  ...  411  412  421  422  423  511            512  521  522  523  
0   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
1   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
2   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
3   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
4   0.0  ...    0  0.0    0    0  0.0  0.0  284728.392578    0    0  0.0  
..  ...  ...  ...  ...  ...  ...  ...  ...            ...  ...  ...  ...  
58  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
59  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
60  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
61  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
62  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  

[63 rows x 45 columns]

In [37]:
for CLC in classCLC:
    by_sites_df[CLC] = by_sites_df.apply(distance_area, args = (CLC,), axis = 1)

In [38]:
by_sites_df

Site  111           112           121         122  123  124  131  132  \
0   F0069  0.0  1.442780e+03  3.587087e+05    0.000000  0.0  0.0  0.0  0.0   
1   F0025  0.0  0.000000e+00  0.000000e+00  700.801763  0.0  0.0  0.0  0.0   
2   F0027  0.0  2.604161e+02  0.000000e+00    0.000000  0.0  0.0  0.0  0.0   
3   B0128  0.0  2.412962e+03  4.053755e+06  523.436355  0.0  0.0  0.0  0.0   
4   F0023  0.0  0.000000e+00  7.907522e+02    0.000000  0.0  0.0  0.0  0.0   
..    ...  ...           ...           ...         ...  ...  ...  ...  ...   
58  F0007  0.0  0.000000e+00  0.000000e+00    0.000000  0.0  0.0  0.0  0.0   
59  B0069  0.0  7.328884e+03  0.000000e+00    0.000000  0.0  0.0  0.0  0.0   
60  B0152  0.0  4.575279e+06  5.227880e+02  421.402885  0.0  0.0  0.0  0.0   
61  F0003  0.0  3.859675e+02  0.000000e+00    0.000000  0.0  0.0  0.0  0.0   
62  B0158  0.0  1.574508e+02  0.000000e+00  392.706036  0.0  0.0  0.0  0.0   

    133  ...  411  412  421  422  423  511         512  521  522  523  
0   0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
1   0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
2   0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
3   0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
4   0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  463.344615  0.0  0.0  0.0  
..  ...  ...  ...  ...  ...  ...  ...  ...         ...  ...  ...  ...  
58  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
59  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
60  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
61  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
62  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  

[63 rows x 45 columns]

check if correct

In [39]:
distance = distances_df[(distances_df['Site'] == 'F0023') & 
                        (distances_df['classCLC'] == 512)]['dist'].values
distance

array([614.50674791])

In [40]:
area = surfaces_df[surfaces_df['Site'] == 'F0023']['512'].values[0]
area

284728.392578125

In [47]:
area/distance[0]

463.3446150839585

### Merging the datasets into two

3 datasets to 2 : sites + (heavy metal || pesticides)

##### Heavy Metal

##### Pesticides

full

In [42]:
pesticides_df.sample(5)

REF....SUBSTANCE   PERIOD  Aldrin  Benalaxyl  Chlorpropham  Cyanofenphos  \
114            B0041  2018/P2     NaN        NaN           NaN           NaN   
262            F0013  2019/P2     NaN        0.0           NaN           NaN   
27             B0039  2017/P2     NaN        0.0           NaN           NaN   
194            B0069  2018/P4     NaN        0.0           NaN           NaN   
148            B0037  2018/P3     NaN        0.0           NaN           NaN   

     Cypermethrin  Ethofumesate  Fludioxonil  Fluopyram  ...  \
114           NaN           NaN          0.0        0.0  ...   
262           NaN           NaN          0.0        0.0  ...   
27            NaN           NaN          NaN        NaN  ...   
194           NaN           NaN          NaN        NaN  ...   
148           NaN           NaN          0.0        NaN  ...   

     qtt_insecticides  qtt_herbicides  nb_fungicides  nb_insecticides  \
114               0.0             0.0              0                0   
262               0.0             0.0              0                0   
27                0.0             0.0              0                0   
194               0.0             0.0              0                0   
148               0.0             0.0              0                0   

     nb_herbicides  lmr_fungicides  lmr_insecticides  lmr_herbicides  year  \
114              0               0                 0               0  2018   
262              0               0                 0               0  2019   
27               0               0                 0               0  2017   
194              0               0                 0               0  2018   
148              0               0                 0               0  2018   

     trimester  
114          2  
262          2  
27           2  
194          4  
148          3  

[5 rows x 128 columns]

In [43]:
small_pesticides_df.sample(5)

REF....SUBSTANCE  qtt_fungicides  qtt_insecticides  qtt_herbicides  \
114            B0041          0.0000               0.0          0.0000   
516            F0063          0.1004               0.0          0.0374   
76             B0017          0.0676               0.0          0.0000   
128            F0007          0.0000               0.0          0.0000   
195            B0017          0.0000               0.0          0.0000   

     nb_fungicides  nb_insecticides  nb_herbicides  lmr_fungicides  \
114              0                0              0               0   
516              1                0              1               1   
76               1                0              0               1   
128              0                0              0               0   
195              0                0              0               0   

     lmr_insecticides  lmr_herbicides  year trimester  
114                 0               0  2018         2  
516                 0               0  2020         4  
76                  0               0  2018         1  
128                 0               0  2018         2  
195                 0               0  2019         1

In [44]:
group_pesti_df.sample(5)

lmr_fungicides  lmr_insecticides  lmr_herbicides
REF....SUBSTANCE                                                  
B0116                          0                 0               0
F0041                          0                 0               0
B0096                          0                 0               0
B0021                          1                 0               0
F0065                          0                 0               0

##### sites

In [45]:
sites_df.sample(5)

Site  classCLC         dist       surface country
179  B0163       111     0.000000  3.145850e+06      BE
52   F0071       311   646.518558  5.607325e+05      FR
372  F0031       231  1140.202604  4.599038e+06      FR
149  F0043       142  1017.359576  2.365331e+05      FR
384  B0023       142  1223.368654  1.945341e+05      BE

In [46]:
by_sites_df.sample(5)

Site          111           112           121          122  123  124  \
0   F0069     0.000000  1.442780e+03  3.587087e+05     0.000000  0.0  0.0   
21  B0140  1416.095618  2.959663e+03  1.258021e+06  1967.770812  0.0  0.0   
57  F0011     0.000000  7.531979e+02  0.000000e+00     0.000000  0.0  0.0   
26  B0039  3098.057557  2.951602e+06  0.000000e+00     0.000000  0.0  0.0   
59  B0069     0.000000  7.328884e+03  0.000000e+00     0.000000  0.0  0.0   

    131  132  133  ...  411  412  421  422         423         511  512  521  \
0   0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0    0.000000    0.000000  0.0  0.0   
21  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0    0.000000  544.152451  0.0  0.0   
57  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0    0.000000    0.000000  0.0  0.0   
26  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0  413.179539    0.000000  0.0  0.0   
59  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0    0.000000    0.000000  0.0  0.0   

    522         523  
0   0.0    0.000000  
21  0.0    0.000000  
57  0.0    0.000000  
26  0.0  776.102211  
59  0.0    0.000000  

[5 rows x 45 columns]

#### Merge

### Changing the NA

# Questions to ask

- When NaN => information irrelevant or not analysed?
- Difference France/Belgium?
- Is the fact that an insecticide is unauthorized relevant? (=> not the same law in France/Belgium)
- Or do we have to show that even banned pesticides are still found?
- How do we treat other categorical pesticides?
- How to increase the number of datas?

### Bo-bo-bonus

- FR  :  36 beehives for 335 zones
- BE  :  27 beehives for 297 zones